### PR TITLE
[WIP] [Parser] Decouple the parser from AST creation (part 3.1)

### PR DESCRIPTION
**Pull request by jansvoboda11**
Instead of creating the AST directly in the parser (and libSyntax or SwiftSyntax via SyntaxParsingContext), make Parser to explicitly create a tree of ParsedSyntaxNodes. Their OpaqueSyntaxNodes can be either libSyntax or SwiftSyntax. If AST is needed, it can be generated from the libSyntax tree.
This PR contains the refactor of a few parsing methods for expressions.

This PR is a part of a Google Summer of Code 2019 [project](https://swift.org/gsoc2019/#integration-of-libsyntax-with-the-rest-of-the-compiler-pipeline).

Work in progress, depends on #26478.

### CODE REVIEW

File: lib/Parse/ASTGen.cpp [hunk 1]
```diff
@@ -10,80 +10,738 @@
...
+  TypeRepr *WrappedType = generate(Type.getWrappedType(), Loc);
+  auto ExclamationLoc = advanceLocBegin(Loc, Type.getExclamationMark());
+  return new (Context)
+      ImplicitlyUnwrappedOptionalTypeRepr(WrappedType, ExclamationLoc);
+}
+
+TypeRepr *ASTGen::generate(UnknownTypeSyntax Type, SourceLoc &Loc) {
+  auto ChildrenCount = Type.getNumChildren();
+
+  if (ChildrenCount == 3 || ChildrenCount == 4) {
```
Comment (rintaro) [inline lib/Parse/ASTGen.cpp]: Could you add comments for each `Unknown*Syntax` conversion branch?
For example this branch could be:
```
// Constract 'ArrayTypeRepr' from obsoleted syntax. i.e.
//     type-simple '[' ...
```

File: lib/Parse/ASTGen.cpp [hunk 2]
```diff
@@ -10,80 +10,738 @@
...
+  if (ChildrenCount == 1) {
+    auto Keyword = Type.getChild(0)->getAs<TokenSyntax>();
+
+    auto IsAtStartOfLine = [](const TokenSyntax Token) {
+      auto LeadingTrivia = Token.getLeadingTrivia();
+      if (LeadingTrivia.empty())
+        return false;
+      auto LastLeadingTriviaKind = Token.getLeadingTrivia().back().getKind();
+      return LastLeadingTriviaKind == TriviaKind::Newline ||
+             LastLeadingTriviaKind == TriviaKind::CarriageReturnLineFeed;
```
Comment (rintaro) [inline lib/Parse/ASTGen.cpp]: This `IsAtStartOfLine` is not quite correct.
```
let abc: /*
*/ class
```
This `class` is "at start of line" but the trivia doesn't contain 'Newline' or `CarriageReturnLineFeed` piece.
Also, can't we just omit this `IsAtStartOfLine` check? The parser already checked `!isAtStartOfLine()`, so if we reach here, I think we can just construct `ErrorTypeRepr` if it's keyword token.

File: lib/Parse/ASTGen.cpp (hunk 1 above)
Comment (rintaro) [reply lib/Parse/ASTGen.cpp]: Ah sorry, these comments are to https://github.com/apple/swift/pull/26478

Comment (jansvoboda11) [reply lib/Parse/ASTGen.cpp]: No worries, I've implemented and pushed both of your suggestions to #26478.

**Comment by rintaro** [comment]
Stale. Closing.

**Pull request closed without merging**